### PR TITLE
Make job_queue database-aware

### PR DESCRIPTION
**Pull request by rishabhs95**
I've tried to keep the functions similar to the Ruby's Queue class to make minimum changes to the code. Issue #320.

The PR is not ready yet but you can see and suggest if there should be any changes. Meanwhile I'll make some changes to make it better and functional with the current API. @claudijd @pwnbus @agaurav77 @royalharsh.

### CODE REVIEW

File: lib/ssh_scan/database/mongo_queue.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        port = opts["port"]
+        socket = server + ":" + port.to_s
+
+        queue_client = Mongo::Client.new([socket], :database => name)
+        return SSHScan::QueueDB::MongoDb.new(queue_client)
+      end
+
+      # @param [String] socket
+      def push(socket)
+        @queue_collection.insert_one("socket" => socket, "status" => "pending")
```
Comment (royalharsh) [inline lib/ssh_scan/database/mongo_queue.rb]: Maybe we can just add an extra field `status` in `ssh_scan.db` and be done away with a second db `ssh_scan_queue.db`?

Comment (rishabhs95) [inline lib/ssh_scan/database/mongo_queue.rb]: We are going to use the database to store scan results anyways, even if we switch job_queue later to something like redis, that is why I wanted to keep them de-coupled.

File: lib/ssh_scan/database/mongo_queue.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
+require 'mongo'
+
+Mongo::Logger.logger.level = ::Logger::FATAL
+
+module SSHScan
+  module QueueDB
+    class MongoDb
+      attr_reader :queue_collection
```
Comment (claudijd) [inline lib/ssh_scan/database/mongo_queue.rb]: I think we'd be better off here if we leveraged the pre-existing DB classes and made JobQueue just use the high-level DB abstraction.

Comment (claudijd) [inline lib/ssh_scan/database/mongo_queue.rb]: SSHScan::JobQueue has methods that talk to SSHScan::DB

-OR-

SSHScan::DB has added methods for queuing and we drop SSHScan::JobQueue entirely

Comment (claudijd) [inline lib/ssh_scan/database/mongo_queue.rb]: My concern with the existing approach is that it created a lot of new code and maybe that's not needed for this additional function.  Really all we're doing is adding an additional table for queue info.

File: lib/ssh_scan/job_queue.rb [hunk 3]
```diff
@@ -1,7 +1,28 @@
...
+    def self.from_hash(opts)
+      database_options = opts["database"]
+
+      # Figure out what database object to load
+      case database_options["type"]
+      when "sqlite"
+        queue_database = SSHScan::QueueDB::SQLite.from_hash(database_options)
+      when "mongodb"
+        queue_database = SSHScan::QueueDB::MongoDb.from_hash(database_options)
+      else
```
Comment (claudijd) [inline lib/ssh_scan/job_queue.rb]: by going the approach I mentioned above, we can skip having to do this again

**Comment by claudijd** [comment]
Would like to propose we moved this PR over to https://github.com/mozilla/ssh_scan_api

Leaving PR open until we have an eqivalent one in the new repo.

**Comment by rishabhs95** [comment]
@claudijd I will move the PR over to `ssh_scan_api` over this week and change the current implementation to a better implementation :)

**Comment by claudijd** [comment]
Please move this PR to the ssh_scan_api project (https://github.com/mozilla/ssh_scan_api)

**Comment by rishabhs95** [comment]
@claudijd I got caught up in some work... will move the PR over to the other repository before this weekend

**Comment by rishabhs95** [comment]
Closing, as this PR has been migrated to `ssh_scan_api`

**Pull request closed without merging**